src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -8,240 +8,108 @@
-      return crimson::ct_error::erange::make();
-    }
-  }
-
-  journal_seq_t j_seq = get_written_to();
-  ceph::bufferlist to_write = encode_record(
-    std::move(record), device->get_block_size(),
-    j_seq, 0);
-  assert(to_write.length() == encoded_size);
-  auto target = get_rbm_addr(get_written_to());
-  auto new_written_to = target + encoded_size;
-  if (new_written_to >= get_journal_end()) {
-    assert(new_written_to == get_journal_end());
-    DEBUG("roll");
-    paddr_t paddr = convert_abs_addr_to_paddr(
-      get_records_start(),
-      get_device_id());
-    set_written_to(
-      journal_seq_t{++written_to.segment_seq, paddr});
-  } else {
-    paddr_t paddr = convert_abs_addr_to_paddr(
-      new_written_to,
-      get_device_id());
-    set_written_to(
-      journal_seq_t{written_to.segment_seq, paddr});
-  }
-  DEBUG("{}, target {}", r_size, target);
-
-  auto write_result = write_result_t{
-    j_seq,
-    encoded_size
-  };
-  auto write_fut = device_write_bl(target, to_write);
-  return handle.enter(write_pipeline->device_submission
-  ).then([write_fut = std::move(write_fut)]() mutable {
-    return std::move(write_fut);
-  }).safe_then([this, &handle] {
-    return handle.enter(write_pipeline->finalize);
-  }).safe_then([this, target,
-    length=encoded_size,
-    write_result,
-    r_size,
-    FNAME] {
-    DEBUG("commit target {} used_size {} written length {}",
-          target, get_records_used_size(), length);
-
-    paddr_t paddr = convert_abs_addr_to_paddr(
-      target + r_size.get_mdlength(),
-      get_device_id());
-    auto submit_result = record_locator_t{
-      paddr,
-      write_result
-    };
-    trimmer.set_journal_head(write_result.start_seq);
-    return submit_result;
-  });
+  return do_submit_record(std::move(record), handle);
 }
 
-CircularBoundedJournal::write_ertr::future<> CircularBoundedJournal::device_write_bl(
-    rbm_abs_addr offset, bufferlist &bl)
+CircularBoundedJournal::submit_record_ret
+CircularBoundedJournal::do_submit_record(
+  record_t &&record,
+  OrderingHandle &handle)
 {
-  LOG_PREFIX(CircularBoundedJournal::device_write_bl);
-  auto length = bl.length();
-  if (offset + length > get_journal_end()) {
-    return crimson::ct_error::erange::make();
+  LOG_PREFIX(CircularBoundedJournal::do_submit_record);
+  if (!record_submitter.is_available()) {
+    DEBUG("H{} wait ...", (void*)&handle);
+    return record_submitter.wait_available(
+    ).safe_then([this, record=std::move(record), &handle]() mutable {
+      return do_submit_record(std::move(record), handle);
+    });
+  }
+  auto action = record_submitter.check_action(record.size);
+  if (action == RecordSubmitter::action_t::ROLL) {
+    return record_submitter.roll_segment(
+    ).safe_then([this, record=std::move(record), &handle]() mutable {
+      // reset committed_to to records_start after roll is done
+      record_submitter.update_committed_to(get_written_to());
```

Comment:
Crosschecking SegmentedJournal, why update_committed_to() needs to be called here?

Rolling doesn't require that the ongoing writes are finished, this is why committed-to was not updated here.

Reply:
This is because current replay in CBJ tries to read records from committed_to in the record. I added the last commit to resolve without this change.

Reply:
OK, committed_to will be used correctly once scan_valid_records() is reused from segmented journal.